ingestion-beam/src/main/java/com/mozilla/telemetry/heka/HekaReader.java

```diff
@@ -0,0 +1,118 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package com.mozilla.telemetry.heka;
+
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessage;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.xerial.snappy.Snappy;
+
+public class HekaReader {
+
+  private static void insertPath(JSONObject o, List<String> s, Object v) {
+    String key = s.get(0);
+    if (s.size() == 1) {
+      o.put(key, v);
+    } else {
+      JSONObject newValue = o.has(key) ? o.getJSONObject(key) : new JSONObject("{}");
+      insertPath(newValue, s.subList(1, s.size()), v);
+      o.put(key, newValue);
+    }
+  }
+
+  private static JSONObject readHekaMessage(InputStream is) {
+    try {
+      while (true) {
+        // continue reading until we find a heka message or we reach the end of the
+        // file
+        // FIXME: add some code to keep track of errors, parseErrors does something
+        // like this
+        int cursor = is.read();
+        if (cursor == -1) {
+          return null;
+        } else if (cursor == 0x1E) {
+          // found seperator, continue
+          break;
+        }
+      }
+
+      int headerLength = is.read();
+      byte[] headerBuffer = new byte[headerLength];
+      is.read(headerBuffer);
+      Heka.Header header = Heka.Header.parseFrom(headerBuffer);
+
+      // Parse unit separator
+      if (is.read() != 0x1F) {
+        throw new IOException("Invalid Heka Frame: missing unit separator");
+      }
+
+      // get message data
+      byte[] messageBuffer = new byte[header.getMessageLength()];
+      is.read(messageBuffer);
+
+      int uncompressedLength = Snappy.uncompressedLength(messageBuffer);
+      byte[] uncompressedMessage = new byte[uncompressedLength];
+      Snappy.uncompress(messageBuffer, 0, header.getMessageLength(), uncompressedMessage, 0);
+      Heka.Message message = Heka.Message.parseFrom(uncompressedMessage);
+
+      JSONObject payload = new JSONObject(new String(message.getPayload().getBytes()));
```

Comment:
We just in the last week decided to isolate usage of `org.json.JSONObject` as much as possible, so you're going to have some errors after rebasing. I didn't anticipate that conflicting with your work, so I apologize for that. I'm happy to hop in and do refactoring to have this use jackson `ObjectNode` instead (using our `Json.asObjectNode(message.getPayload())`).

Reply:
No problem, I don't think this should be too difficult to deal with. I think I will try to get things working at least locally with the current structure, then do a rebase.